File: Legacy/Leetcode/92ReverseLinkedListII.py
# Reverse a linked list from position m to n. Do it in-place and in one-pass.
#
# For example:
# Given 1->2->3->4->5->NULL, m = 2 and n = 4,
#
# return 1->4->3->2->5->NULL.
#
# Note:
# Given m, n satisfy the following condition:
# 1 ≤ m ≤ n ≤ length of list.
#

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        if m == n:
            return head
        dummy = ListNode(0)
        dummy.next = head
        previous = dummy
        for i in range(m - 1):
            previous = previous.next
        start = previous.next
        first = previous.next
        second = first.next
        for i in range(n - m):
            temp = ListNode(second.val)
            temp.next = second.next
            second.next = first
            first = second
            second = temp.next
        previous.next = first
        start.next = second
        return dummy.next

File: C++/Linked List/92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* s = new ListNode(0, head);
        ListNode* p1 = s, * p2, * p3, * p4;
        for (int i = 0; i < left-1; i++) p1 = p1->next;
        p2 = p1->next;
        p3 = p2; p4 = p2->next;
        for (int i = 0; i < right - left; i++) {
            ListNode* temp = p4->next;
            p4->next = p3;
            p3 = p4;
            p4 = temp;
        }
        p1->next = p3;
        p2->next = p4;
        return s->next;
    }
};